packages/runtime/container-runtime/src/id-compressor/README.md

```diff
@@ -1,3 +1,69 @@
+- Prior to the service being notified of the ID's creation, it is unique within the scope of the session (e.g. it can collide with IDs allocated by a different allocator in the same network).
+- After the service has been notified of the ID's creation and has determined a total ordering between all clients creating IDs, the ID will have a _final_ form that is a small number representation that is unique across the [document](#document).
+
+## **ID Spaces**
+
+The allocation scheme separates the small-number IDs into two "spaces":
+
+1. **Session space**: IDs are unique within the session scope, but are not guaranteed to be [document](#document) unique without the accompanying context of their session ID.
+2. **Op space**: IDs are in their most final form, as unique as possible within the entire system. IDs that have been ordered by the central service ("finalized") are represented in their _final_ form, while any other IDs that have not yet been finalized are left in their _session space_ form.
+
+Each of these spaces is represented by a discrete type. This allows for the direct encoding of an ID's uniqueness guarantees without contextual ambiguity. An ID can be "normalized" from one space to the other with sychronous call.
+
+## **Usage Guidelines**
+
+Allocated IDs can be delivered to application authors for use in place of V4 UUIDs, with the following considerations:
+
+- Expose only session space IDs to application authors. This means that application developers don't need to manage multiple ID spaces or worry about converting IDs between session space and op space. This simplifies the data management process and reduces the chances of errors or inconsistencies.
+
+  - **Note:** Session space IDs are only unique within the originating session (for most applications this will mean within each individual client's app instance, as it is rare to hold handles to more than one allocator). This constrains usage to scopes where IDs are unique; for instance, application developers should never persist these IDs and should instead be directed to convert them to StableIds first.
+
+- Use op space IDs for serialized forms, such as sending information over the wire or storing it in a file. Op space IDs are in their _final_ form if possible, which imparts a performance benefit by reducing normalization work when a client receives a network operation or rehydrates an allocator.
+- When serializing op space IDs, annotate the entire context (e.g., file or network operation) with the session ID. This is necessary because not all IDs in op space are in their _final_ form. Any non-finalized IDs will still be in their session space form, unique only within that specific session. Annotating the entire context with the session ID ensures that recipients of the serialized data can correctly interpret and process the non-finalized IDs.
```

Comment:
Websocket: "Wow, -1, you're so small and lightweight."
Negative Seventeen: "_This isn't even my final form_!"